Microsoft.NET.Build.Containers/ContainerHelpers.cs

```diff
@@ -96,42 +60,102 @@ public static bool IsValidImageName(string imageName)
+    /// This is primarily for testing - in the real world almost all usage should be through HTTPS!
+    /// </summary>
+    public static Uri TryExpandRegistryToUri(string alreadyValidatedDomain)
+    {
+        var prefix = alreadyValidatedDomain.StartsWith("localhost") ? "http" : "https";
+        return new Uri($"{prefix}://{alreadyValidatedDomain}");
     }
 
     /// <summary>
     /// Parse a fully qualified container name (e.g. https://mcr.microsoft.com/dotnet/runtime:6.0)
     /// Note: Tag not required.
     /// </summary>
+    /// <remarks>
+    /// This code is adapted from <see href="https://github.com/distribution/distribution/blob/78b9c98c5c31c30d74f9acb7d96f98552f2cf78f/reference/reference.go#L191-L236">reference.go</see>
+    /// and so may not match .NET idioms.
+    /// NOTE: We explicitly are not handling digest references at the moment.
+    /// </remarks>
     /// <param name="fullyQualifiedContainerName"></param>
     /// <param name="containerRegistry"></param>
     /// <param name="containerName"></param>
     /// <param name="containerTag"></param>
+    /// <param name="containerDigest"></param>
     /// <returns>True if the parse was successful. When false is returned, all out vars are set to empty strings.</returns>
     public static bool TryParseFullyQualifiedContainerName(string fullyQualifiedContainerName,
                                                             [NotNullWhen(true)] out string? containerRegistry,
                                                             [NotNullWhen(true)] out string? containerName,
-                                                            [NotNullWhen(true)] out string? containerTag)
+                                                            out string? containerTag, // tag is always optional - we can't guarantee anything here
+                                                            out string? containerDigest // digest is always optional - we can't guarantee anything here
+                                                            )
     {
-        Uri? uri = ContainerImageToUri(fullyQualifiedContainerName);
 
-        if (uri == null || uri.Segments.Length <= 1)
+        /// if we don't have a reference at all, bail out
+        var referenceMatch = ReferenceRegexp.Match(fullyQualifiedContainerName);
+        if (referenceMatch is not { Success: true })
+        {
+            containerRegistry = null;
+            containerName = null;
+            containerTag = null;
+            containerDigest = null;
+            return false;
+        }
+
+        // if we have a reference, then we have three groups:
+        // * reference name
+        // * reference tag (optional)
+        // * reference digest (optional)
+
+        // this will always be successful if the ReferenceRegexp matched, so it's safe to index into.
+        var namePortion = referenceMatch.Groups[1].Value;
```

Comment:
What does index 0 tend to be in these groups?

Reply:
for all regex matches, index 0 is the entire matched string, not any of the separate capture groups